README.md
# About PDS

Package Distro Search tool (abbreviated as PDS), allows to perform package search across multiple OS distributions in a single UI interface.


# Documentation

Documentation like how to add extra distro support are available in the docs directory [here](docs/Adding_new_distros.md). 

# Installation

Instructions to install can be found in the docs directory [here](docs/Installation.md).

# License

Package Distro Search tool (abbreviated as PDS) is licensed under the [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0).

# Contribution

Documentation to explains how to contribute changes to the PDS project is available in the docs directory.
Contributor License Agreement(CLA) for both individual and corporate are available in the docs directory.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## v1.5.42
## Enchancements
-- Updated Data files
-- Added rhel 8.6

## v1.5.41
## Enchancements
-- Updated Data files
-- Added Ubuntu 22.04

## v1.5.40
## Enchancements
-- Updated Data files
-- Removed Ubuntu 21.04

## v1.5.39
## Enchancements
-- Updated Data files

## v1.5.38
## Enchancements
-- Updated Data files

## v1.5.37
## Enchancements
-- Updated Data files

## v1.5.36
## Enchancements
-- Updated Data files
-- Added rhel 8.5

## v1.5.35
## Enchancements
-- Updated Data files
-- Added ubuntu 21.10

## v1.5.34
## Enchancements
-- Updated Data files
-- Added minor UI fix

## v1.5.33
## Enchancements
-- Updated Data files

## v1.5.32
## Enchancements
-- Updated Data files
-- Added sles 15.3
-- Removed Ubuntu 20.10

## v1.5.31
## Enchancements
-- Updated Data files
-- Added rhel 8.4

## v1.5.30
## Enchancements
-- Updated Data files
-- Added UI fix, Updated setup instructions

## v1.5.29
## Enchancements
-- Updated Data files
-- Added ub 21.04

## v1.5.28
## Enchancements
-- Updated Data files
-- Added back ub 16.04
-- Added back rhel 7.6,7.7,6.10,8.0
-- Added back sles 12 SP3,SP4,15

## v1.5.27
## Enchancements
-- Updated Data files

## v1.5.26
## Enchancements
-- Updated Data files

## v1.5.25
## Enchancements
-- Updated Data files

## v1.5.24
## Enchancements
-- Updated Data files
-- Added RHEL 8.3
-- Removed RHEL 7.7 ,Sles 15

## v1.5.23
## Enchancements
-- Updated Data files
-- Added Ubuntu 20.10, RHEL 7.9
-- Removed Ubuntu 16.04, RHEL 7.6

## v1.5.22
## Enchancements
-- Updated Data files

## v1.5.21
## Enchancements
-- Updated Data files
-- Added SLES 15-SP2

## v1.5.20
## Enchancements
-- Updated Data files

## v1.5.19
## Enchancements
-- Updated Data files
-- Removed Ubuntu 19.10,SLES 12 SP4

## v1.5.18
## Enchancements
-- Updated Data files


## v1.5.17
## Enchancements
-- Updated Data files
-- Added RHEL 8.2, Ubuntu 20.04
-- Removed RHEL 6.10/7.4/7.5
-- Added minor search fix

## v1.5.16
## Enchancements
- Updated supported Data files
- Added RHEL 7.8

## v1.5.15
## Enchancements
- Updated supported Data files
- Added SLES 12-SP5
- removed ubuntu 19.04 

## v1.5.14
## Enchancements
- Added UI fix - change the search limit from 3 to 2 characters.

## v1.5.13
## Enchancements
- Updated supported Data files
- Added rhel 8.1

## v1.5.12
## Enchancements
- Updated supported Data files

## v1.5.11
## Enchancements
- Updated supported Data files
- Updated Links in FAQ page

## v1.5.10
## Enchancements
- Updated supported Data files
- Added ubuntu 19.10
- Added rhel 

## v1.5.9
## Enchancements
- Updated supported Data files 

## v1.5.8
## Enchancements
- Updated supported Data files

## v1.5.7
## Enchancements
- Updated supported Data files
- Added SLES 15 SP1 support

## v1.5.6
## Enchancements
- Updated supported Data files

## v1.5.5
## Enchancements
- Updated supported Data files

## v1.5.4
## Enchancements
- Added Ub 19.04
- Updated supported Data files

## v1.5.3
## Enchancements
- Added necessary data files
- Updated versions of Sles 12 and Alpine

## v1.5.2
## Enchancements
- Added necessary data files

## v1.5.1
## Enchancements
- Added necessary data files
- Commented intel section of code

## v1.5
## Enchancements
- Updated tool with added support to searches on intel
- Added necessary data files
- UI changes to provide search options for intel based distributions

## v1.4.6
## Enchancements
- Updated existing data files

## v1.4.5
## Enchancements
- Added SLES 15
- Updated existing data files

## v1.4.4
### Enhancements
- Added Ub 18.04
- Updated supported Data files

## v1.4.3
### Enhancements
- Updated Ubuntu data files.

## v1.4.2
### Enhancements
- Updated FAQ with new SLES distro.
- Updated data files
- Updated SUSE Package HUB with support for Sles12-SP3
- Removed support for Ubuntu 16.10
- Added Support for Ubuntu 17.10

## v1.4
### Features
- Responsive UI - Displays well on wide range of display devices including mobile devices

### Enhancements
- New UI to provide results refinement, provide additional details about package availability by distro version
- New UI displays FAQ in accordion view
- Fetch results in background of results are too much
- Updated search API signiture to reflect correct paramters
- PDS can now also be accessed on root (e.g. http://<server>[:port]/ ) instead of /pds route.
- Now using 'Response' module to set response headers correctly
- Removed 'Disclaimer' controller

## v1.3
### Features:
- Added SUSE Package Hub support
- Distro data files with package lists refreshed on 2017-June-19
### Enhancements:
- Search packages case-insensitive
- Added total page count on pagination control
- Simplified 'Adding New Distro' functionality along with document
- Updated FAQ

## v1.2
### Features:
- Adding Ubuntu 17.04 support
- Performance optimizations to improve search functionality.


src/config/config.py
import urllib.request, urllib.error, urllib.parse
import logging

PDS_BASE = '/opt/PDS/'
DATA_FILE_LOCATION = '%s/distro_data' % PDS_BASE
LOG_FILE_LOCATION = '%s/log/pds.log' % PDS_BASE
MIN_DATA_FILE_SIZE = 50000
MAX_RECORDS_TO_CONCAT = 5000
MAX_RECORDS_TO_SEND = 100
CACHE_SIZE = 10
STATS_FILE_LOCATION = '/opt/PDS/stats'
proxy_user = 'proxy_user'
proxy_password = 'proxy_password'
proxy_server = 'proxy_server'
proxy_port = 'proxy_port'

enable_proxy_authentication = False
server_host = '0.0.0.0'
server_port = 5000

DISABLE_PAGINATION = False
STATS_SECRET='new123T'

'''
Following are the various debug levels:
CRITICAL
ERROR
WARNING
INFO
DEBUG
NOTSET
Refer https://docs.python.org/2/library/logging.html for more information.
'''
DEBUG_LEVEL = logging.ERROR

SUPPORTED_DISTROS = {
   "zUbuntu":{
      "zUbuntu 16.04":"Ubuntu_16_04_Package_List.json",
      "zUbuntu 18.04":"Ubuntu_18_04_Package_List.json",
      "zUbuntu 20.04":"Ubuntu_20_04_Package_List.json",
      "zUbuntu 21.10":"Ubuntu_21_10_Package_List.json",
      "zUbuntu 22.04":"Ubuntu_22_04_Package_List.json"
   },
   "zSUSE Linux Enterprise Server":{
      "zSLES 12 SP3":"Suse_Linux_Enterprise_Server_12_SP3_Package_List.json",
      "zSLES 12 SP4":"Suse_Linux_Enterprise_Server_12_SP4_Package_List.json",
      "zSLES 12 SP5":"Suse_Linux_Enterprise_Server_12_SP5_Package_List.json",
      "zSLES 15":"Suse_Linux_Enterprise_Server_15_Package_List.json",
      "zSLES 15 SP1":"Suse_Linux_Enterprise_Server_15_SP1_Package_List.json",
      "zSLES 15 SP2":"Suse_Linux_Enterprise_Server_15_SP2_Package_List.json",
      "zSLES 15 SP3":"Suse_Linux_Enterprise_Server_15_SP3_Package_List.json"
   },
   "zSUSE Package Hub":{
      "zSLES 12 SP3":"SUSE_Package_Hub_SLES_12_SP3.json",
      "zSLES 12 SP4":"SUSE_Package_Hub_SLES_12_SP4.json",
      "zSLES 12 SP5":"SUSE_Package_Hub_SLES_12_SP5.json",
      "zSLES 15":"SUSE_Package_Hub_SLES_15.json",
      "zSLES 15 SP1":"SUSE_Package_Hub_SLES_15_SP1.json",
      "zSLES 15 SP2":"SUSE_Package_Hub_SLES_15_SP2.json",
      "zSLES 15 SP3":"SUSE_Package_Hub_SLES_15_SP3.json"
   },
   "zRHEL":{
      "zRHEL 6.10":"RHEL_6_10_Package_List.json",
      "zRHEL 7.6":"RHEL_7_6_Package_List.json",
      "zRHEL 7.7":"RHEL_7_7_Package_List.json",
      "zRHEL 7.8":"RHEL_7_8_Package_List.json",
      "zRHEL 7.9":"RHEL_7_9_Package_List.json",
      "zRHEL 8.0":"RHEL_8_Package_List.json",
      "zRHEL 8.1":"RHEL_8_1_Package_List.json",
      "zRHEL 8.2":"RHEL_8_2_Package_List.json",
      "zRHEL 8.3":"RHEL_8_3_Package_List.json",
      "zRHEL 8.4":"RHEL_8_4_Package_List.json",
      "zRHEL 8.5":"RHEL_8_5_Package_List.json",
      "zRHEL 8.6":"RHEL_8_6_Package_List.json"
   },
   "xUbuntu":{
      "xUbuntu 16.04":"xUbuntu_16_04_Package_List.json",
      "xUbuntu 18.04":"xUbuntu_18_04_Package_List.json",
      "xUbuntu 20.04":"xUbuntu_20_04_Package_List.json",
      "xUbuntu 21.10":"xUbuntu_21_10_Package_List.json",
      "xUbuntu 22.04":"xUbuntu_22_04_Package_List.json"
   },
   "xSUSE Linux Enterprise Server":{
      "xSLES 12 SP3":"xSuse_Linux_Enterprise_Server_12_SP3_Package_List.json",
      "xSLES 12 SP4":"xSuse_Linux_Enterprise_Server_12_SP4_Package_List.json",
      "xSLES 12 SP5":"xSuse_Linux_Enterprise_Server_12_SP5_Package_List.json",
      "xSLES 15":"xSuse_Linux_Enterprise_Server_15_Package_List.json",
      "xSLES 15 SP1":"xSuse_Linux_Enterprise_Server_15_SP1_Package_List.json",
      "xSLES 15 SP2":"xSuse_Linux_Enterprise_Server_15_SP2_Package_List.json",
      "xSLES 15 SP3":"xSuse_Linux_Enterprise_Server_15_SP3_Package_List.json"
   },
   "xSUSE Package Hub":{
      "xSLES 12 SP3":"xSUSE_Package_Hub_SLES_12_SP3.json",
      "xSLES 12 SP4":"xSUSE_Package_Hub_SLES_12_SP4.json",
      "xSLES 12 SP5":"xSUSE_Package_Hub_SLES_12_SP5.json",
      "xSLES 15":"xSUSE_Package_Hub_SLES_15.json",
      "xSLES 15 SP1":"xSUSE_Package_Hub_SLES_15_SP1.json",
      "xSLES 15 SP2":"xSUSE_Package_Hub_SLES_15_SP2.json",
      "xSLES 15 SP3":"xSUSE_Package_Hub_SLES_15_SP3.json"
   },
   "xRHEL":{
      "xRHEL 6.10":"xRHEL_6_10_Package_List.json",
      "xRHEL 7.6":"xRHEL_7_6_Package_List.json",
      "xRHEL 7.7":"xRHEL_7_7_Package_List.json",
      "xRHEL 7.8":"xRHEL_7_8_Package_List.json",
      "xRHEL 7.9":"xRHEL_7_9_Package_List.json",
      "xRHEL 8.0":"xRHEL_8_Package_List.json",
      "xRHEL 8.1":"xRHEL_8_1_Package_List.json",
      "xRHEL 8.2":"xRHEL_8_2_Package_List.json",
      "xRHEL 8.3":"xRHEL_8_3_Package_List.json",
      "xRHEL 8.4":"xRHEL_8_4_Package_List.json",
      "xRHEL 8.5":"xRHEL_8_5_Package_List.json",
      "xRHEL 8.6":"xRHEL_8_6_Package_List.json"
   }
}

logging.basicConfig(format='%(asctime)s %(message)s', filename=LOG_FILE_LOCATION, level=DEBUG_LEVEL)

LOGGER = logging.getLogger('PDS_SERVER')
   
# In case application is hosted on server with proxy, set "enable_proxy_authentication = True" in config.py 
# and update the proxy details
def proxy_authentication():
    proxy = urllib.request.ProxyHandler({'http': 'http://%s:%s@%s:%s' % (proxy_user, proxy_password, proxy_server, proxy_port),
      'https': 'https://%s:%s@%s:%s' % (proxy_user, proxy_password, proxy_server, proxy_port)}
    )
    auth = urllib.request.HTTPBasicAuthHandler()
    opener = urllib.request.build_opener(proxy, auth, urllib.request.HTTPHandler)
    urllib.request.install_opener(opener)
